CHI-35: Add clustering of free text terms to ISARIC analytics

ed-phillips opened the pull request:
Feature branch to add processing and visualization code to support clustering of free text terms in Vertex.

ed-phillips commented:
@tedinburgh @EstebanGarciaG here's a draft PR to get started on adding the freetext clustering to Vertex. I've added a couple of functions to produce a dataframe for the first visualization we talked about - the 'intertopic distance map', showing how big the topics are and how close together they sit in embedding space.

Whenever you get time please weigh in on:
- if the `cluster_df` object as described is indeed suitable for our needs right now
- the processing functions we need to get the input for `get_clusters`, that is a list of strings comprising the free-text terms in a given field

tedinburgh commented:
@ed-phillips thanks for this. I think the input for `get_clusters` works fine as a list (or a pandas series), and the preprocessing would be handled within the script for an insight panel to allow it to be applied to different types of free text variables. For the most part (with the current data schema), I think big proportion of the free text cells would be empty, so the only additional step would be to remove missing/NaN/empty strings? I am also not sure whether there are cases in which a multivariate topic modelling would be better than modelling each variable independently, in which case it might be worth considering allowing the dataframe to be an input?

tedinburgh commented:
My final question is about potentially having some optional keyword arguments in the `get_clusters` to allow for (limited) user specification of the topic modelling behaviour (which is something you've mentioned in the comments). So I think it would be useful to know what are the most important parameters/hyperparameters to BERTopic, from the perspective of an end user who isn't an expert in this (e.g. number of topics). (For context, we probably expect most local (software developer) users to create or adapt an insight panel, but probably fewer users will go deep into IsaricAnalytics to tweak hyperparameters within the function.)

ed-phillips commented:
Thanks for the comments @tedinburgh 

- I've added some cleaning of the input list of strings
- I'm not sure I fully understand what you mean by multivariate topic modelling - in any case it'd probably require slightly different methodology than exists in the current function. I'd suggest keeping this function as is and adding a `get_multivariate_clusters` fn or similar down the line, depending on what we'd want from that function. Open to other opinions though
- I've added `nr_topics` as an argument to `get_clusters`. I'm reluctant to add anything else as an arg, because while number of topics is fairly intutive, I think you'd want to know what you were doing if experimenting with embedding or representation models, i.e. it's not so easy for a casual developer to play around with
